lib/rack/query_parser.rb

```diff
@@ -215,20 +211,37 @@ def params_hash_has_key?(hash, key)
       true
     end
 
-    def check_query_string(qs, sep)
-      if qs
-        if qs.bytesize > @bytesize_limit
-          raise QueryLimitError, "total query size (#{qs.bytesize}) exceeds limit (#{@bytesize_limit})"
-        end
+    def each_query_pair(qs, separator, unescaper = nil)
+      return enum_for(:each_query_pair, qs, separator, unescaper) unless block_given?
 
-        if (param_count = qs.count(sep.is_a?(String) ? sep : '&')) >= @params_limit
-          raise QueryLimitError, "total number of query parameters (#{param_count+1}) exceeds limit (#{@params_limit})"
-        end
+      return if !qs || qs.empty?
 
-        qs
+      if qs.bytesize > @bytesize_limit
+        raise QueryLimitError, "total query size (#{qs.bytesize}) exceeds limit (#{@bytesize_limit})"
+      end
+
+      pairs = qs.split(separator ? (COMMON_SEP[separator] || /[#{separator}] */n) : DEFAULT_SEP, @params_limit + 1)
```

Comment:
As this had become the only caller, I've inlined `check_query_string` (probably uncontroversial), and in doing so also modified how it works (maybe more controversial?), by integrating the parameter count check with the `split`. Happy to unwind if we're concerned it might be less safe.

Reply:
This makes the common case more efficient, with the trade-off of increased work in the pathological case.  However, as we only split up to the limit, I agree this is a better approach.